### PR TITLE
Only check ExternalDependency when customizing kotlin dependencies

### PR DESCRIPTION
**Pull request by bingranl**
This commit fixes the issue where kgp checks group/version of project
dependencies which is uncompatible with project isolation. With this
change, kgp will only check ExternalDependency which should be enough
for setting up kotlin dependencies.

This change also changes the way we apply cache redirector gradle
files to all projects when setting up test projects. To make it
compatible with project isolation, we should move away from using
project.getAllProjects() function.

Bug: KT-47792
Test: existing + SimpleKotlinGradleIT.testProjectIsolation

### CODE REVIEW

**Comment by bingranl** [comment]
Hey @Tapchicoma, this PR is ready for review and all the tests should pass. Thanks!

**Review by h0tk3y — APPROVED** [review]

File: libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/KotlinDependenciesManagement.kt [hunk 1]
```diff
@@ -192,7 +192,7 @@ private fun chooseAndAddStdlibDependency(
 
         // If the exact same module is added in the source sets hierarchy, possibly even with a different scope, we don't add it
         val moduleAddedInHierarchy = hierarchySourceSetsDependencyConfigurations.any {
-            it.allDependencies.any { dependency -> dependency.group == KOTLIN_MODULE_GROUP && dependency.name == stdlibModuleName }
+            it.allExternalDependencies().any { dependency -> dependency.group == KOTLIN_MODULE_GROUP && dependency.name == stdlibModuleName }
```
Comment (Tapchicoma) [inline libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/KotlinDependenciesManagement.kt:195]: According to Gradle folks:
> you can iterate over `allDependencies` and query any of the dependencies. However, `ProjectDependency.getGroup()`  and `ProjectDependency.getVersion()` will fail as these access the state of the target project.

Current approach excludes `FileCollectionDependency` and while adding stdlib jar is low probability, it still may happen. How about filter out only `ProjectDependency`s?
```kt
it.allDependencies.matching { it !is ProjectDependency }
```

File: libraries/tools/kotlin-gradle-plugin-integration-tests/src/test/kotlin/org/jetbrains/kotlin/gradle/testbase/testDsl.kt [hunk 2]
```diff
@@ -231,30 +231,33 @@ internal fun Path.enableCacheRedirector() {
...
-                """
-                
-                allprojects {
-                    apply(from = "${'$'}rootDir/gradle/cacheRedirector.gradle.kts")
-                }
-                
-                """.trimIndent()
-            )
+    val projectDir = toFile()
+    projectDir.walk().forEach {
```
Comment (Tapchicoma) [inline libraries/tools/kotlin-gradle-plugin-integration-tests/src/test/kotlin/org/jetbrains/kotlin/gradle/testbase/testDsl.kt:235]: Why such change is required? `allprojects {}` is not compatible with project isolation? :thinking:

Comment (bingranl) [reply libraries/tools/kotlin-gradle-plugin-integration-tests/src/test/kotlin/org/jetbrains/kotlin/gradle/testbase/testDsl.kt:235]: Yup, it is not compatible cause you are accessing subproject from root project. See https://gradle.github.io/configuration-cache/

File: libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/KotlinDependenciesManagement.kt (hunk 1 above)
Comment (bingranl) [reply libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/KotlinDependenciesManagement.kt]: Done

**Review by Tapchicoma — APPROVED** [review]

**Comment by Tapchicoma** [comment]
Merged via cherry-pick into `master`: https://github.com/JetBrains/kotlin/commit/c57302abbab998c1ce73c31fb4b71eb1c150d1fe

**Pull request closed without merging**